ocaml/vhd-tool/src/impl.ml

```diff
@@ -829,6 +829,31 @@ let retry common retries f =
   in
   aux retries
 
+let stream_with_relative ~common ~nbd_dev ~relative_to =
+  ( match Image.of_device nbd_dev with
+  | Some (`Vhd vhd) ->
+      return vhd
+  | _ ->
+      fail (Failure "Not found a VHD image behind the NBD device")
+  )
+  >>= fun vhd ->
+  let path = common.path @ [Filename.dirname vhd] in
+  retry common 3 (fun () -> Vhd_IO.openchain ~path vhd false) >>= fun t ->
+  Vhd_IO.close t >>= fun () ->
+  Vhd_format_lwt.IO.openfile nbd_dev false >>= fun raw ->
+  Vhd_IO.openchain ~path relative_to false >>= fun t' ->
+  Vhd_IO.close t' >>= fun () -> Hybrid_input.vhd ?from:(Some t') raw t
```

Comment:
Here `nbdhybrid` was requested at the `source_format` of the stream, which implies using just the NBD sources to read the data using the `Nbd_input` module (and therefore this works for both vhd and qcow). The "hybrid" part is that it uses the "raw" source served by `nbd-client` and separately gets information about the sparseness of the file from the NBD server directly.

But now this ends up getting translated to `Hybrid_input` if the `relative_to` parameter is set, which belongs to the `hybrid` source format, and directly looks at the VHD files behind the scenes (and therefore only works for VHDs). I don't like implicitly switching from one mode to the other. If we are going this route, then we should get xapi to ask for `hybrid` right away, rather than `nbdhybrid`.

A better way for the future, however, would be to use NBD even for the base source, but this may not be possible yet. I think we are missing the same sort of functionality that we need for storage migration using NBD in this case.

CC: @MarkSymsCtx

Reply:
To look at the simple case firstly without considering the `relative_to`, I understand:
1. `Hybrid_input.vhd raw vhd1 reads from the `raw`, gets sparseness info from the `vhd`, and generates VHD stream.
2. `Nbd_input.vhd raw nbd_server export_name size` reads the `raw`, gets sparseness info from the `nbd_server`, etc, and generates a VHD stream.

The differences are on:
a. how to get the sparseness info, and
b. `Hybrid_input` can get sparseness from the backend in VHD format, but `Nbd_input` can get sparseness from others in a format like QCOW2 hidden by the NBD interface.

Now considering the `relative_to`, I understand the use case applies only to the VHD source and VHD base/relative_to. So additionally, we need to compare the blocks:
a. between the `block from raw` and `block from vhd` in `relative_to`.   
b.  between the `block from raw` and `block from NBD` in `relative_to`.

Reply:
@MarkSymsCtx pointed a couple of tickets to use NBD interface as the source directly. Then the comparison would be between the `block from NBD` in `source` and `block from NBD` in r`elative_to`.

Reply:
> then we should get xapi to ask for hybrid right away
This is a good point. Now I realized that this `hybrid` doesn't have to be the Frankentap device only. This also agrees with the use case which applies only to the VHD source and VHD base/relative_to.

Reply:
>Now considering the relative_to, I understand the use case applies only to the VHD source and VHD base/relative_to.

We have this restriction at the moment. But conceptually, difference disks are not unique to the VHD format. I think the future will be NBD for everything, and something lower-level to handle different formats (VHD, QCOW2, whatever).